e2e/testcases/presync_test.go

```diff
@@ -0,0 +1,265 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/policy"
+	"kpt.dev/configsync/e2e/nomostest/registryproviders"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/e2e/nomostest/testwatcher"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/core/k8sobjects"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reposync"
+	"kpt.dev/configsync/pkg/rootsync"
+	"kpt.dev/configsync/pkg/status"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var cosignPassword = map[string]string{
+	"COSIGN_PASSWORD": "test",
+}
+
+func TestAddPreSyncAnnotationRepoSync(t *testing.T) {
+	repoSyncID := core.RepoSyncID(configsync.RepoSyncName, namespaceRepo)
+	nt := nomostest.New(t, nomostesting.SyncSource,
+		ntopts.RequireOCIProvider,
+		ntopts.SyncWithGitSource(repoSyncID),
+		ntopts.RepoSyncPermissions(policy.RBACAdmin(), policy.CoreAdmin()))
+
+	if err := nomostest.SetupOCISignatureVerification(nt); err != nil {
+		nt.T.Fatal(err)
+	}
+	rootSyncGitRepo := nt.SyncSourceGitReadWriteRepository(nomostest.DefaultRootSyncID)
+	repoSyncKey := repoSyncID.ObjectKey
+	gitSource := nt.SyncSources[repoSyncID]
+
+	nt.T.Log("Create first OCI image with latest tag.")
+	bookInfoRole := k8sobjects.RoleObject(core.Name("book-info-admin"))
+	image, err := nt.BuildAndPushOCIImage(repoSyncKey, registryproviders.ImageInputObjects(nt.Scheme, bookInfoRole))
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Create RepoSync with first OCI image")
+	repoSyncOCI := nt.RepoSyncObjectOCI(repoSyncKey, image.OCIImageID().WithoutDigest(), "", image.Digest)
+	nt.Must(rootSyncGitRepo.Add(nomostest.StructuredNSPath(repoSyncID.Namespace, repoSyncID.Name), repoSyncOCI))
+	nt.Must(rootSyncGitRepo.CommitAndPush("Set the RepoSync to sync from OCI"))
+	nt.Must(nt.Watcher.WatchObject(kinds.RepoSyncV1Beta1(), repoSyncID.Name, namespaceRepo,
+		testwatcher.WatchPredicates(testpredicates.RepoSyncHasSourceError(status.SourceErrorCode, "no signatures found"))))
+
+	nt.T.Log("Create second image with latest tag.")
+	bookstoreSA := k8sobjects.ServiceAccountObject("bookstore-sa")
+	image1, err := nt.BuildAndPushOCIImage(repoSyncKey, registryproviders.ImageInputObjects(nt.Scheme, bookInfoRole, bookstoreSA))
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err = signImage(nt, image); err != nil {
```

Comment:
We don't verify the first image is synced successfully. Do we need to sign it?

Reply:
We want to make sure even though the first image is signed, when the latest tag is with second unsigned image, the server would keep blocking the request - to prevent a false positive.

Reply:
ack